### ISSUE TITLE
Transformer and CPU path with `src_mask` raises error with torch 1.12

### ISSUE THREAD

**Issue by ajfadam** [pytorch/pytorch#81129]
### 🐛 Describe the bug

The following code, which runs on torch 1.11 cpu, doesn't anymore on torch 1.12:
```python
import torch

model = torch.nn.TransformerEncoderLayer(d_model=512, nhead=8, batch_first=True)
src = torch.rand(32, 10, 512)
src_mask = torch.zeros(10, 10).to(torch.bool)

model.eval()
with torch.no_grad():
    print(model(src, src_mask))
```
It raises 
``` 
Traceback (most recent call last):
  File "/Users/adm/Desktop/main.py", line 9, in <module>
    print(model(src, src_mask))
  File "/Users/adm/pytorch/torch/nn/modules/module.py", line 1186, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/adm/pytorch/torch/nn/modules/transformer.py", line 439, in forward
    return torch._transformer_encoder_layer_fwd(
RuntimeError: Expected attn_mask->sizes()[0] == batch_size to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

The `.to(torch.bool)` is only here to silence the warning: `UserWarning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly.`, the original code on v1.11 doesn't use it.

This also happens on an x86 ubuntu machine, when using cpu, but it does not happens when using CUDA.

Because of this condition: https://github.com/pytorch/pytorch/blob/e9b3bc2eadb8ffe10c002abcd5a34a5b7d36f390/aten/src/ATen/native/transformers/attention.cpp#L136-L144
using `src_mask = torch.zeros(32, 10)` makes the error go away, but it must not be right because I believe the size of `src_mask` should be `(seq_len, seq_len)`. 
Instead, on torch 1.11, using `(32, 10)` was raising `RuntimeError: The shape of the 2D attn_mask is torch.Size([32, 10]), but should be (10, 10).`. 
Also, the comment saying CPU path doesn't support mask makes me wonder if I'm looking at the right code 🤔 

### Versions

Collecting environment information...
PyTorch version: 1.13.0a0+git4c57cf9
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:07:06) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] torch==1.13.0a0+git4c57cf9
[conda] numpy                     1.23.0          py310h0a343b5_0    conda-forge
[conda] torch                     1.13.0a0+git4c57cf9           dev_0    <develop>

cc @ezyang @gchanan @zou3519 @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by erichan1** [comment]
This is a new transformer implementation issue. Thank you for bringing this up - we'll get a fix asap.

**Comment by erichan1** [comment]
Relevant function is 
https://github.com/pytorch/pytorch/blob/e9b3bc2eadb8ffe10c002abcd5a34a5b7d36f390/aten/src/ATen/native/transformers/attention.cpp#L118-L150

Tagging @swolchok. Looks like cpu path assumed src_key_padding_mask with shape (batch size, seqlen) and not src_mask with shape (seqlen, seqlen)? src_mask works fine on the cuda path since _masked_softmax takes care of it.

**Comment by swolchok** [comment]
IMO, best solution is probably to have parity between the CPU and CUDA paths in masked_softmax w.r.t. "transformer mask" support, then we don't have to continue layering more special cases onto the special-case fix for the gap in support.

**Comment by erichan1** [comment]
Main PR to fix is https://github.com/pytorch/pytorch/pull/81277. I've also taken the liberty to increase our test coverage in test/test_transformers.py. For cherry-picking onto the 1.12.1 release we will also need https://github.com/pytorch/pytorch/pull/79796 because it created test_transformers.py but didn't get into the 1.12 release. cc @atalman

**Comment by erichan1** [comment]
@ajfadam Fix was #81277 which landed yesterday. Should work now - let me know if issues remain.

**Comment by ajfadam** [comment]
Thanks for the quick fix !
